src/pip/_internal/self_outdated_check.py

```diff
@@ -96,6 +100,102 @@ def was_installed_by_pip(pkg: str) -> bool:
     return dist is not None and "pip" == dist.installer
 
 
+def get_py_executable() -> str:
+    """Get path to launch a Python executable.
+
+    First test if python/python3/pythonX.Y on PATH matches the current
+    interpreter, and use that if possible. Then try to get the correct
+    pylauncher command to launch a process of the current python
+    version, fallback to sys.executable
+    """
+
+    if not sys.executable:
+        # docs (python 3.10) says that sys.executable can be can be None or an
+        # empty string if this value cannot be determined, although this is
+        # very rare. In this case, there is nothing much we can do
+        return "python3"
+
+    # windows paths are case-insensitive, pathlib takes that into account
+    sys_executable_path = Path(sys.executable)
+
+    major, minor, *_ = sys.version_info
+
+    # first handle common case: test if path to python/python3/pythonX.Y
+    # matches sys.executable
+    for py in ("python", "python3", f"python{major}.{minor}"):
+        which = shutil.which(py)
+        if which is None:
+            continue
+
+        try:
+            # resolve() removes symlinks, normalises paths and makes them
+            # absolute
+            if Path(which).resolve() == sys_executable_path.resolve():
+                return py
```

Comment:
What if the user has a shell alias that overrides what `shutil.which` returns? Then what we suggest won't work as we expect it to.

Also on a case-insensitive filesystem, this `==` needs to be a case insensitive comparison (although this will simply miss some possible abbreviations, not return something wrong).

Reply:
In both these cases won't this check return `False` then and correctly use the `sys.executable` fallback?